src/commands/Changelog.ts

```diff
@@ -0,0 +1,318 @@
+                    csvSeparator +
+                    "Added to language(s):" +
+                    csvSeparator +
+                    "Removed from language(s):" +
+                    csvSeparator +
+                    "Remains in language(s):" +
+                    csvSeparator +
+                    "Used for calculating the metric(s):" +
+                    csvSeparator +
+                    "Was explicitly only activated for language(s):" +
+                    EOL
+            );
+
+            for (const entry of this.getAll().values()) {
+                if (
+                    entry.remainingLanguages.size > 0 ||
+                    (entry.addedLanguages.size > 0 && entry.removedLanguages.size > 0)
+                ) {
+                    const mapping = metricMappings.get(entry.expression);
+                    if (mapping !== undefined) {
+                        const activatedForLangOutput =
+                            mapping.activated_for_languages === undefined
+                                ? ""
+                                : mapping.activated_for_languages;
+                        writeStream.write(
+                            escapeForCsv(entry.expression) +
+                                csvSeparator +
+                                Array.from(entry.addedLanguages) +
+                                csvSeparator +
+                                Array.from(entry.removedLanguages) +
+                                csvSeparator +
+                                Array.from(entry.remainingLanguages) +
+                                csvSeparator +
+                                mapping.metrics +
+                                csvSeparator +
+                                activatedForLangOutput +
+                                EOL
+                        );
+                    } else {
+                        throw new Error(
+                            "Programming mistake: No existing expression mapping for a syntax node that is not new: " +
+                                entry.expression
+                        );
+                    }
+                }
+            }
+            writeStream.end();
+        });
+    }
+}
+
+/**
+ * Tracks the changes that were applied to one expression.
+ */
+class ChangelogEntry {
+    /**
+     * Name of the expression.
+     */
+    expression: string;
+
+    /**
+     * Whether this is a totally new node that was not included in the mappings file before.
+     */
+    isNew: boolean;
+
+    /**
+     * Languages that included this node type before.
+     */
+    oldLanguages: Set<string>;
+
+    /**
+     * Languages from which this node type was removed.
+     */
+    removedLanguages: Set<string>;
+
+    /**
+     * Languages to which this node type was added.
+     */
+    addedLanguages: Set<string>;
+
+    /**
+     * Languages that include this node type and had included it before.
+     */
+    remainingLanguages: Set<string>;
```

Comment:
It is not necessary to store remainingLanguages as a property here. It can be derived from old, removed and addedLanguages properties. providing a getter like `remainingLanguages()` would not force you to inject the needed data it in the constructor.

Reply:
Mh, it would be oldLanguages - removedLanguages, basically. However, hat would force me to calculate the remaining languages each time I call that getter, and as remainingLanguages is used more than once, that would increase the computing effort. (This [Set difference operation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set/difference) would be elegant in this case, but it does not seem to be available for anything but Safari yet 😉)

Injecting the data to the constructor should not be an issue as it is only performed by the Changelog-class inside the method `removedNodeFromLanguage` and sets it to the same value as `oldLanguages` there.

However, I notice that I do never use `oldLanguages`, so I could remove that filed from the ChangelogEntry. 
Even if one decides to include that in the CSV-file in the future, this would only be a one-time merge between remaining and removed languages (or just re-introduce the field oldLanguages).